BUG: fix reference count error on invalid input to ndarray.flat

adeak opened the pull request:
According to https://github.com/numpy/numpy/issues/13165 there's a reference count error when invalid inputs are given to `ndarray.flat`. The fix is to remove a superfluous `Py_DECREF` call in the invalid input case.

My reasoning for the fix is that [here](https://github.com/numpy/numpy/blob/251c526672f897543176b966229ecde96bb89a9d/numpy/core/src/multiarray/iterators.c#L630) we get a new reference to the new dtype as `indtype`, but in case of blatantly invalid input (say, `None` or a string) we only enter [this block](https://github.com/numpy/numpy/blob/251c526672f897543176b966229ecde96bb89a9d/numpy/core/src/multiarray/iterators.c#L638-L641) and then next [we decrease the refcount of `indtype`](https://github.com/numpy/numpy/blob/251c526672f897543176b966229ecde96bb89a9d/numpy/core/src/multiarray/iterators.c#L669-L671). I believe the bug was caused by the double deallocation together with the final `Py_XDECREF` call at the end. The `else` branch I removed was only tripped in cases where the input was invalid (otherwise we either return from the corresponding `if` branch or jump to the `fail` label from there). In every valid case we return with either [calling `Py_DECREF(indtype)` exactly once](https://github.com/numpy/numpy/blob/251c526672f897543176b966229ecde96bb89a9d/numpy/core/src/multiarray/iterators.c#L647) or [passing on its reference](https://github.com/numpy/numpy/blob/251c526672f897543176b966229ecde96bb89a9d/numpy/core/src/multiarray/iterators.c#L652-L653), and in non-trivially-invalid cases [we used to skip the now-removed `else` branch](https://github.com/numpy/numpy/blob/251c526672f897543176b966229ecde96bb89a9d/numpy/core/src/multiarray/iterators.c#L663-L665).

I intended to add a regression test [as suggested by @seberg](https://github.com/numpy/numpy/issues/13165#issuecomment-475157997), but the bug causes a reference count error to be printed after the interpreter finishes. I don't know how I could test this (the `IndexError` is raised fine so a test would conclude just fine).

Closes https://github.com/numpy/numpy/issues/13165.

adeak committed:
BUG: fix reference count error on invalid input to ndarray.flat

seberg commented:
@adeak you may be able to check the reference count e.g. on an input dtype of `np.intp` inside the tests. I think I would still like a reference test if it does not get ugly, even if before the test the error only shows up when exiting the interpreter (e.g. we may add reference count tests at some point). Most likely that error path is missing a test in any case. If it is not easy to make a test that explicitly fails on the old version, that is fine.

However, I think you should be able to check `sys.getrefcount(np.dtype(np.intp))`, maybe check `git grep "pytest.*HAS_REFCOUNT"` because such a test has to be disabled on pypy.

adeak commented:
I did try doing that, but I did it wrong: I tried checking the refcount of `np.intp` rather than `np.dtype(np.intp)` (and that didn't change whether or not the bug was present). I also wasn't sure how good form it was to call `sys.getrefcount` in a test. Thanks for the tip, I'll investigate it.

adeak commented:
@seberg awesome, getting the refcount of `np.dtype(np.intp)` does indeed work, thank you! I'll put together a test with the pypy consideration.

adeak committed:
TST: add refcount check for ndarray.flat

Primarily a regression test for issue gh-13165, but also check
the refcount of the input and valid use cases while we're at it.

adeak commented:
Added a test. I decided to check the refcount of the input as well, since there are multiple changes to its refcount throughout the code. I also added valid use cases just to cover as many paths as possible.

I noticed that all the other refcount-related tests in `test_multiarray.py` use direct `if HAS_REFCOUNT` tests rather than the pytest decorator (even the few tests which are entirely about refcounts), but I decided to stick with the decorator that's present in a lot of the other tests.

mattip commented:
The other refcount tests predate pytest. Using a decorator is nicer, at some point we may want to refactor the other tests.

adeak commented:
Exactly on of the Travis tests is failing. It's for Python 3.6 with the following envvars:
```
USE_WHEEL=1 RUN_FULL_TESTS=1 RUN_COVERAGE=1 INSTALL_PICKLE5=1
```
There's another test with `INSTALL_PICKLE5` and one with `USE_WHEEL` that passed, and I'd expect `RUN_FULL_TESTS` not to be relevant, because I suspect that affects "slow" tests. Any chance that `RUN_COVERAGE` introduces additional references somehow? Or does my test make wrong assumptions?

The assertion that's tripped is that the refcount of the input index itself doesn't change on call.

seberg commented:
@adeak if your test fails try running manually `gc.collect()` before checking refcounts. Alternatively, repeat the tests many times and check that the reference count does not change too much. Testing reference counts can be tricky.

adeak commented:
@seberg I see, thanks, that explains the fuzzy "refcount didn't change more than 50 times in a 100-length loop" checks elsewhere in the test file. I'll do that first, then.

One technical question: should I amend my commit and force-push, or should I just add new commits? The former is bad because force pushing, the latter will lead to messy history unless the commits get squashed on merge.

seberg commented:
@adeak just do what you prefer. Forced pushing to a PR branch is not really considered bad, usually it is just you really working on it anyway. But we can just squash it into a single commit anyway and github now has a button to do that, so there is not much reason for you to worry about it.

adeak committed:
TST: add fuzzying to ndarray.flat refcount test for robustness

mhvk commented on numpy/core/src/multiarray/iterators.c:
```diff
@@ -666,9 +666,6 @@ iter_subscript(PyArrayIterObject *self, PyObject *ind)
         Py_DECREF(obj);
         return (PyObject *)ret;
     }
-    else {
```
This is very nicely found!

I would prefer, though, to correct it a bit differently: change the `if` statement on line 643 to:
```
if (!PyArray_Check(obj)) {
    goto fail;
}
```
and then un-indent what is inside the current `if` clause. That way, the "good" part clearly ends with a `return` before the `fail` branch. What do you think?

adeak replied:
Of course, you're right, that looks much cleaner, thank you. But then I think I'll slightly refactor that block further, the disconnect between the return on line 666 and its direct predecessor on line 647 stuck out to me when I first looked at it.

adeak committed:
MAINT: refactor array input case in ndarray.flat for clarity

eric-wieser commented on numpy/core/src/multiarray/iterators.c:
```diff
@@ -640,34 +640,32 @@ iter_subscript(PyArrayIterObject *self, PyObject *ind)
+        new = PyArray_FromAny(obj, indtype, 0, 0,
+                          NPY_ARRAY_FORCECAST | NPY_ARRAY_ALIGNED, NULL);
+        if (new == NULL) {
+            goto fail;
```
This causes a double decreed, the reference to indtype was stolen on line 659

adeak replied:
Indeed, you're right! Just to be clear, this was there [in the original too](https://github.com/numpy/numpy/blob/251c526672f897543176b966229ecde96bb89a9d/numpy/core/src/multiarray/iterators.c#L651-L655), right? I did have an uncertain feeling about that, but I was still confused about some bits of the code when I noticed that, and then I forgot.

The question is, can it happen that we have an integer array but the conversion on (the new) line 659 fails? That's the only case when this problem arises, and I should add that as a test case. But in order for this to be foolproof I'll have to increment the refcount, and decrement it if we succeeded. Thank you for spotting it!

eric-wieser replied:
I think the ref should be stolen unconditionally, whether `new` succeeds or not. So your latest revision still double-decrefs `indtype`

adeak replied:
I'm probably missing something still. But as I see it, we [start with one reference to `indtype`](https://github.com/numpy/numpy/blob/64445bcc9ec8926f828bc109d103eb0f0d001e38/numpy/core/src/multiarray/iterators.c#L630-L633), then [bump up the refcount of `indtype`](https://github.com/numpy/numpy/blob/64445bcc9ec8926f828bc109d103eb0f0d001e38/numpy/core/src/multiarray/iterators.c#L659) because `PyArray_FromAny` steals one, so we'll still have one to spare after the call. Then either the call succeeds and we get rid of that spare reference manually before returning,
```
        Py_INCREF(indtype);
        new = PyArray_FromAny(obj, indtype, 0, 0,
                          NPY_ARRAY_FORCECAST | NPY_ARRAY_ALIGNED, NULL);
        if (new == NULL) {
            goto fail;
        }
        Py_DECREF(indtype);
        /* ... */
        return new;
```
or we jump to `fail` and let `XDECREF` do it:
```
 fail:
    /* ... */
    Py_XDECREF(indtype);
```
So in both cases we should have [from line 659](https://github.com/numpy/numpy/blob/64445bcc9ec8926f828bc109d103eb0f0d001e38/numpy/core/src/multiarray/iterators.c#L659) two references to `indtype`, one of which is stolen by `...FromAny` and one which is disposed of manually. Or, conversely, we only need to bump the refcount of `indtype` before the call to `...FromAny` because the `fail` mode includes a call to `XDECREF`.

eric-wieser replied:
You're right, I missed the incref

adeak committed:
BUG: remove additional double decref bug from ndarray.flat

mhvk commented on numpy/core/src/multiarray/iterators.c:
```diff
@@ -640,34 +640,34 @@ iter_subscript(PyArrayIterObject *self, PyObject *ind)
-    else {
+
+    /* Check for integer array */
+    if (PyArray_ISINTEGER((PyArrayObject *)obj)) {
```
Can we still invert this? I.e., 
```
if (!PyArray_ISINTEGER(...) {
    goto fail;
}
```
That way we don't use `fail` as a fall-through at all.

adeak replied:
I don't think that's equivalent. The `else` of this if is a no-op. The only remaining fall-through is on line 662 when the conversion to an array failed. Similar to the fall-through on line 634. Am I missing something?

adeak replied:
I mean, we can invert that `if` and move the contents of the original `if` block up one indent, but we'd still have that one `goto fail` left. I thought your remark was in order to remove the latter.

mhvk replied:
No, that's not what I meant. Right now, you have *here*
```
if (PyArra...) {
  ...
  (goto fail or return something)
}

fail:
```

I'd like it to be
```
if (!PyArray_IS...) {
   goto fail;
}

...
(either goto fail or return something)

(No way to get here except via goto)
fail:
```

Just the same as for the original `if` statement!

mhvk replied:
Though I guess the one issue is that you are creating `new` inside here, which is not possible if we unindent it.  So, perhaps just let it be...

adeak replied:
Ah, I see! I did understand what you were suggesting, but I thought that would be equivalent to what we have now. I somehow managed to forget about the (implicitly no-op) `else` branch of that `if` block... I'm sorry, you're right of course.

Is there any reason why we can't just move the declaration of `new` up next to that of `obj`, for instance? It took me a while to figure out why merely unindenting would not be possible (i.e. that we need to support ANSI C which restricts where declarations can be put) so I don't trust my judgment very much at this hour :)

mhvk replied:
@adeak - indeed, one can just define `new` at the top - which some don't like to the point of adding clauses without even an `if` statement and others do not mind at all. Since you are doing the work, I think you should pick what you prefer! Though we should not let this distract from the fact that you solved the bug! So, if you'd rather just have us have a final look so that this can be merged, that's totally OK!

adeak replied:
Perfect, thanks for the explanation, @mhvk. I did realize the diabolical option to add a superfluous clause. Since the three options seem to be "early declaration", "superfluous clause" and "fallthrough into `fail`", I find the first to be the least offensive. And if by any chance it's decided that my choice is worse after all, whoever merges can just cherry-pick out that commit and keep the current state.

adeak committed:
MAINT: reverse integer array testing logic in ndarray.flat to remove fall-through into fail

eric-wieser commented on numpy/core/src/multiarray/iterators.c:
```diff
@@ -640,35 +641,36 @@ iter_subscript(PyArrayIterObject *self, PyObject *ind)
+    Py_DECREF(obj);
+    obj = new;
+    new = iter_subscript_int(self, (PyArrayObject *)obj);
+    Py_DECREF(obj);
```
These three lines can be spelt `Py_SETREF(new, iter_subscript_int(self, (PyArrayObject *)new))`

mhvk replied:
It may make sense to use `ret` here instead of `new`, so that one returns `ret`. (Clearly we've gotten to the stage of totla nitpicking!)

adeak replied:
I'm OK with improving the code as much as possible while we're at it. Plus it's not my original code we're nitpicking here ;) I'll take both of those suggestions, thanks.

adeak replied:
@mhvk I've held back the renaming for now, because beyond the name `ret` is a `PyArrayObject *` and I'm not sure how I could cast safely. Furthermore, we're going to return a `PyObject *`, so it feels a bit weird to cast back and forth in two subsequent lines.

On a related note, is it not weird that some of the functions in `iterators.c` return `PyObject *`s while others `PyArrayObject *`s? In particular, one out of three `iter_subscript_*` functions returns the latter. If there's a sane way to use `ret` after all we can change it of course, I just don't know how.

mhvk replied:
@adeak - definitely fine to keep as is. I notice that somewhere about `ret` is used in exactly the poor way you describe - but let's leave that out for now!

adeak committed:
MAINT: use Py_SETREF in ndarray.flat

adeak commented:
I can't tell if the [failing test](https://dev.azure.com/numpy/numpy/_build/results?buildId=2573) is related. 32-bit 3.6 passed, 64-bit 3.7 passed, 32-bit 3.7 failed on Windows. `matmul` shouldn't end up doing anything with `ndarray.flat`, right?

seberg commented:
@adeak I believe that is a random failure that sometimes occurs (possibly related to a bad OpenBLAS version on the test machine). You can safely ignore it.

adeak commented:
Pushing a no-op change to trigger a rerun of the tests.

mhvk reviewed (approved):
Looks all OK to me. @eric-wieser - do you want another look? If not, maybe just go ahead and merge (squashing commits).

eric-wieser commented:
Looks fine to me too. I think @mhvk was right, reusing `ret` might have been clearer than using `SETREF`, but really not worth another update for. Thanks for the fix!

adeak commented:
@eric-wieser ah, I missed Marten's point to use `ret` _instead of_ `SETREF`. Having them both is what I couldn't get to do right without the compiler shouting at me, which is why I said I didn't know how to cast safely. Now I tried what I think should be done (`new` -> `iter_subcript_int` into `ret` -> `DECREF` `new`)  and it seems to work now...but I understand if it's not worth another PR.
Thank you for the merge!

eric-wieser commented:
I missed his point the first time too. It's your time, if you want to make a follow up PR I'm happy to spend the 30s review to merge it, but I merged as is to avoid the bug fix getting delayed further.

adeak commented:
OK, thank you, posted https://github.com/numpy/numpy/pull/13243.
